### PR TITLE
much ado

### LINKED ISSUE #7: How to index with term weights produced by neural models (e.g., uniCOIL, Splade)?
Hi,
Thanks a lot for this very useful work. 
I wonder if there is anyway to index collections with pre-computed term weights (similar to [this](https://github.com/castorini/pyserini/blob/master/docs/experiments-unicoil.md) in `pyserini`)?

### LINKED ISSUE #9: avoid numpy in c api
Some of the build complexity comes from the usage of numpy in the C API code (e.g., the need to use skbuild). It also makes the package very picky about which numpy version is installed.

The reason numpy is used is to (1) avoid some GIL overheads during retrieval and (2) speed up the construction of the result dataframe. We may be able to get both of these benefits while dropping the numpy requirement by using the standard library [`array`](https://docs.python.org/3/library/array.html) package instead. A few quick tests suggest that building dataframe columns from `array` objects instead of numpy arrays is faster too.

`array` will work for the rank (int) and score (float) columns, but not for qid or docno columns, sine it does not support variable-length strings. One solution here would be to make use of the [`struct`](https://docs.python.org/3/library/struct.html) package to pack qids/docnos, but that only supports bytes objects, not strings, so we'd just end up needing to decode the bytes in python (likely adding a bunch of overhead). Maybe python lists are not so bad here, given there's already be the overhead of creating string objects (GIL'd) anyway.

### PR DESCRIPTION
**Pull request by seanmacavaney**
- Fixes #9 by passing numpy arrays as buffers instead of returning them. This makes it work far better across numpy versions
 - Fixes #7, #8, by allowing the user to index doc dicts and pass query dicts, thereby allowing for easy indexing and retrieval of models like EPIC, SPLADE, etc.

### CODE REVIEW

File: tests/test_ciff.py [hunk 1]
```diff
@@ -8,7 +8,7 @@ class CiffTest(TempDirTestCase):
     def test_to_and_from_ciff(self):
         from pyterrier_pisa import PisaIndex
         dataset = pt.get_dataset("vaswani")
-        idx = PisaIndex(self.test_dir)
+        idx = PisaIndex(self.test_dir, threads=1)
```
Comment (cmacdonald) [inline tests/test_ciff.py]: why do we now need threads=1 everywhere. Shouldnt threads=1 be default?

File: tests/test_dict.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      import nltk
+      nltk.download('punkt')
+      dataset = pt.get_dataset('irds:vaswani')
+      idx = PisaIndex(self.test_dir+'/index', text_field='text_dict', stemmer='none', threads=1)
+      idx_pipe = DictTokeniser() >> idx
+      idx_pipe.index(dataset.get_corpus_iter())
+      self.assertTrue(idx.built())
+      self.assertEqual(len(idx), 11429)
+      MAPS=[0.2256]
+      TRANS=[DictTokeniser('query') >> pt.apply.rename({'query_dict' : 'query_toks'}) >> idx.bm25()]
```
Comment (cmacdonald) [inline tests/test_dict.py]: Can we standardise on query_toks as the column name?

File: src/pyterrier_pisa/_pisathon.cpp [hunk 3]
```diff
@@ -358,20 +354,27 @@ static PyObject *py_retrieve(PyObject *self, PyObject *args, PyObject *kwargs) {
...
+  Py_buffer result_docnos;
+  Py_buffer result_ranks;
+  Py_buffer result_scores;
+
   /* Parse arguments */
-  static const char *kwlist[] = {"index_dir", "encoding", "algorithm", "scorer_name", "stemmer", "queries", "block_size", "quantize", "bm25_k1", "bm25_b", "pl2_c", "qld_mu", "k", "stop_fname", "threads", NULL};
-  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "sssssO|KIffffIsI", const_cast<char **>(kwlist),
-                                     &index_dir, &encoding, &algorithm, &scorer_name, &stemmer, &in_queries, &block_size, &in_quantize, &bm25_k1, &bm25_b, &pl2_c, &qld_mu, &k, &stop_fname, &threads))
+  static const char *kwlist[] = {"index_dir", "encoding", "algorithm", "scorer_name", "stemmer", "queries", "block_size", "quantize", "bm25_k1", "bm25_b", "pl2_c", "qld_mu", "k", "stop_fname", "threads", "pretokenised", "query_weighted", "result_qidxs", "result_docnos", "result_ranks", "result_scores", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "sssssO|KIffffIsIiiw*w*w*w*", const_cast<char **>(kwlist),//TODO update parse
```
Comment (cmacdonald) [inline src/pyterrier_pisa/_pisathon.cpp]: TODO remaining? 

BTW, can we try to comment this strange list of characters?

**Review by cmacdonald — CHANGES_REQUESTED** [review]
Should we also have a README or example about how to use SPLADE etc.

File: tests/test_ciff.py (hunk 1 above)
Comment (seanmacavaney) [reply tests/test_ciff.py]: Fixed, that was for testing for failing tests on github actions. It turns out it wasn't the issue, but forgot to remove.

threads=1 might be a better default, but there are some pros and cons. Wrote up as a separate issue.

**Comment by cmacdonald** [comment]
Are we good Sean? Have we tested with SPLADE?

**Comment by seanmacavaney** [comment]
Running now!

**Pull request merged**